File: ElgamalEllipticCurve.cpp
#include<iostream>
#include<math.h>
#include<cstdlib>
#include<vector>
using namespace std;

int main()
{
	int n,p;
	cout<<"ElGamal Elliptic Curve Cryptography by \nName - DHRUV DIXIT\nRegister Number - 15BCE1324\nVIT University, Chennai\n\n";
	cout<<"Elliptic Curve General Form \t y^2 mod p = (x^3  + A*x + B) mod p \n";
	cout<<"Enter the value of P: \n";
	cin>>p;
	n = p;
	int LHS[2][n],RHS[2][n],a,b,i,j;

	cout<<"\nEnter the Value of a: \n";
	cin>>a;
	cout<<"\nEnter the Value of b: \n";
	cin>>b;
	cout<<"\nCurrent Elliptic Curve \t\t ---> y^2 mod "<<p<<" = (x^3  + "<<a<<"*x + "<<b<<") mod p\n\n\n";

	vector <int> arr_x;
	vector <int> arr_y;

// Equating LHS and RHS as per arbitrary index to generate range of values.
	for(int i = 0; i <	 n; i++)
	{
		LHS[0][i] = i;
		RHS[0][i] = i;
		LHS[1][i] = ((i*i*i) + a*i + b) % p;
		RHS[1][i] = (i*i) % p;
	}


// Generating Base Points
int in_c = 0;
	for(i = 0; i < n; i ++)
	{

		for(j = 0; j < n; j++)
		{
			if(LHS[1][i] == RHS[1][j])
			{
						in_c++;
					arr_x.push_back(LHS[0][i]);
					arr_y.push_back(RHS[0][j]);
			}
		}
	}

	cout<<endl<<"Generated Points are:"<<endl;

	for(i =0; i < in_c; i++)
	{
		cout<<i+1<<"\t( "<<arr_x[i]<<" , "<<arr_y[i]<<" )"<<endl;
	}
	cout<<"Base Point: ("<<arr_x[0]<<","<<arr_y[0]<<")"<<"\n";
	int k,d,M;
	cout<<"Enter the random number 'd' i.e. Private key of Sender  (d<n)\n";
	cin>>d;
	int Qx=d*arr_x[0];
	int Qy=d*arr_y[0];
	//Q is the public key of sender


	//Encryption
    cout<<"Enter the random number 'k' (k<n)\n";
	cin>>k;

    cout<<"Enter the message to be sent:\n";
	cin>>M;
	cout<<"The message to be sent is:\n"<<M<<"\n";

	int c1x=k*arr_x[0];
	int c1y=k*arr_y[0];
	cout<<"Value of C1: ("<<c1x<<","<<c1y<<")"<<"\n";

	int c2x=k*Qx+M;
	int c2y=k*Qy+M;
	cout<<"Value of C2: ("<<c2x<<","<<c2y<<")"<<"\n";

	//Decryption
	cout<<"\nThe message received is:\n";
	int Mx=c2x-d*c1x;
	int My=c2y-d*c1y;
	cout<<Mx;
}

File: ElgamalEllipticCurve.py
print("ElGamal based Elliptic Curve Cryptography\nby Dhruv Dixit:\t 15BCE1324\nVIT University, Chennai\n\nElliptic Curve General Form:\t y^2 mod n=(x^3  + a*x + b)mod n\nEnter 'n':")

def polynomial(LHS,RHS,n):
    for i in range(0,n):
        LHS[0].append(i)
        RHS[0].append(i)
        LHS[1].append((i*i*i + a*i + b)%n)
        RHS[1].append((i*i)%n)


def points_generate(arr_x,arr_y,n):
    count=0
    for i in range(0,n):
        for j in range(0,n):
            if(LHS[1][i]==RHS[1][j]):
                count+=1
                arr_x.append(LHS[0][i])
                arr_y.append(RHS[0][j])
    return count

#main
n=int(input())
LHS=[[]]
RHS=[[]]
LHS.append([])
RHS.append([])
print("Enter value of 'a':")
a=int(input())
print("Enter value of 'b':")
b=int(input())

#Polynomial
polynomial(LHS,RHS,n)

arr_x=[]
arr_y=[]
#Generating base points
count=points_generate(arr_x,arr_y,n)
    
#Print Generated Points
print("Generated points are:")
for i in range(0,count):
    print(i+1," (",arr_x[i],",",arr_y[i],")\n")



#Calculation of Base Point
bx=arr_x[0]
by=arr_y[0]
print("Base Point taken is:\t(",bx,",",by,")\n")


print("Enter the random number 'd' i.e. Private key of Sender (d<n):")
d=int(input())
if(d>=n):
    print("'d' should be less than 'n'.")
else:
    #Q i.e. sender's public key generation
    Qx=d*bx
    Qy=d*by
    print("Public key of sender is:\t(",Qx,",",Qy,")\n")

    #Encrytion
    print("Enter the random number 'k' (k<n):\n")
    k=int(input())
    if(k>=n):
        print("'k' should be less than 'n'")
    else:
        print("Enter the message to be sent:\n")
        M=int(input())

        #Cipher text 1 generation
        C1x=k*bx
        C1y=k*by
        print("Value of Cipher text 1 i.e. C1:\t(",C1x,",",C1y,")\n")

        #Cipher text 2 generation
        C2x=k*Qx+M
        C2y=k*Qy+M
        print("Value of Cipher text 2 i.e. C2:\t(",C2x,",",C2y,")\n")

    #Decryption
        Mx=C2x-d*C1x
        My=C2y-d*C1y
        print("The message recieved by reciever is:\t",Mx)            
